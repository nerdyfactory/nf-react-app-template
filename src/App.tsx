import React, { useState, useReducer, useEffect } from 'react';
import { ThemeProvider } from '@material-ui/styles/';
import { BrowserRouter } from 'react-router-dom';
import { Toaster } from 'react-hot-toast';

import { PrivateNav, PublicNav } from 'navigations';
import useBootstrap from 'hooks/useBootstrap';
import { AppContext, appReducer, INITIAL_APP_STATE } from 'contexts/AppContext';
import { AuthActionTypes, AuthContext, authReducer, INITIAL_AUTH_STATE } from 'contexts';
import { Loading } from 'components';
import './App.css';
import { USER_TOKEN } from 'constants/utility';
import { getIntialState } from 'services/utils/persist-state';

function App() {
  const [authCtx, authDispatcher] = useReducer(authReducer, INITIAL_AUTH_STATE);
  const [appCtx, appDispatcher] = useReducer(appReducer, INITIAL_APP_STATE);
  const [theme] = useBootstrap();
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    const token = getIntialState(USER_TOKEN);
    if (token) {
      authDispatcher({ type: AuthActionTypes.SET_TOKEN, payload: token });
    }
    setIsLoaded(true);
  }, []);

  if (!isLoaded) return <Loading />;

  return (
    <ThemeProvider theme={theme}>
      <BrowserRouter basename="/">
        <AuthContext.Provider value={{ ...authCtx, dispatch: authDispatcher }}>
          <AppContext.Provider value={{ ...appCtx, dispatch: appDispatcher }}>
            {authCtx.token ? <PrivateNav /> : <PublicNav />}
            <Toaster
              position="top-right"
              toastOptions={{
                style: {
                  margin: '40px',
                  background: '#363636',
                  color: '#fff',
                  zIndex: 99999999,
                },
              }}
            />
          </AppContext.Provider>
        </AuthContext.Provider>
      </BrowserRouter>
    </ThemeProvider>
  );
}

export default App;
