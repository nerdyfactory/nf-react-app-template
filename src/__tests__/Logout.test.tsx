import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import 'jest-localstorage-mock';
import App from '../App';

jest.mock('../services/api');

const VALID_EMAIL = 'test@test.com';
const VALID_PASSWORD = 'test1234';

describe('Logout', () => {
  const loginAfterRender = async () => {
    const res = render(<App />);
    fireEvent.input(screen.getByPlaceholderText('User'), {
      target: {
        value: VALID_EMAIL,
      },
    });
    fireEvent.input(screen.getByPlaceholderText('Password'), {
      target: {
        value: VALID_PASSWORD,
      },
    });
    fireEvent.click(screen.getByText('Login'));
    await waitFor(() => screen.getByText('Login'));
    return res;
  };

  it('renders logout', async () => {
    const { container } = await loginAfterRender();
    expect(container).toMatchSnapshot();
  });

  it('moves to login screen', async () => {
    await loginAfterRender();
    fireEvent.click(screen.getByText('Logout'));
    expect(screen.getAllByRole('button')[0]).toHaveTextContent('Login');
    expect(localStorage.setItem).toHaveBeenCalledTimes(1);
    expect(localStorage.removeItem).toHaveBeenCalledTimes(1);
  });
});
