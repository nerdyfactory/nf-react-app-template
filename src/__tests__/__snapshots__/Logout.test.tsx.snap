// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout renders logout 1`] = `
<div>
  <div
    class="sc-dlfnbm iuHaXN"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-bdfBwQ lfqaYk MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Admin console
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-bdfBwQ lfqaYk MuiButton-containedSecondary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Logout
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    style="position: fixed; z-index: 9999;"
  />
</div>
`;
